NEXUS-20698: Add *.rds file support for R proxy

chernikovmaksym opened the pull request:
https://issues.sonatype.org/browse/NEXUS-20698

Added support for .rds metadata files in proxy and group repos as requested in this issue.

Now installing old versions of packages from archive works as it should

aornatovskyy commented on nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RRecipeSupport.groovy:
```diff
@@ -110,16 +110,27 @@ abstract class RRecipeSupport
+   * Matcher for packages.gz mapping.
    */
-  static Builder packagesMatcher() {
+  static Builder packagesGzMatcher() {
```
can we do here as we have for YUM? https://github.com/sonatype/nexus-internal/blob/bd3ff99c3b28f6d21b1d8b930f66e50d2440ba06/private/plugins/nexus-repository-yum/src/main/java/org/sonatype/nexus/repository/yum/internal/YumRecipeSupport.groovy#L177

aornatovskyy replied:
I mean combine  new TokenMatcher('/{path:.+}/PACKAGES.gz') and  new TokenMatcher('/{path:.+}/{filename:.+}.rds')

chernikovmaksym replied:
Nice idea, done in f5a0c4341c6ed3aa9669cb5cd973d162f75a6acd

sergeyonofreychuck commented on nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RProxyFacetImpl.java:
```diff
@@ -72,12 +72,12 @@ protected void doValidate(final Configuration configuration) throws Exception {
+      case METADATA:
+        return getAsset(extractFullPath(context));
       case ARCHIVE:
         return getAsset(path(path(matcherState), filename(matcherState)));
```
the logic related to path looks too complicated. Please create a technical debt task to refactor this. It looks like a good task for an improvement day

chernikovmaksym replied:
Agree with you. Here's the ticket https://issues.sonatype.org/browse/NEXUS-21202

sergeyonofreychuck commented on nexus-repository-r/src/test/java/org/sonatype/nexus/repository/r/internal/AssetKindTest.java:
```diff
@@ -22,14 +22,13 @@
   public void cacheTypes() throws Exception {
     assertThat(ARCHIVE.getCacheType(), is(equalTo(CONTENT)));
-    assertThat(PACKAGES.getCacheType(), is(equalTo(METADATA)));
+    assertThat(AssetKind.METADATA.getCacheType(), is(equalTo(METADATA)));
```
import AssetKind.METADATA

chernikovmaksym replied:
Done in f5a0c4341c6ed3aa9669cb5cd973d162f75a6acd

chernikovmaksym commented:
> Question: Why have we changed from `PACKAGES` to METADATA`? Do we need to consider existing databases that will already be marked incorrectly as `PACKAGES`?

Before this pull request R format supported only `PACKAGES.gz` metadata. But there are plenty of other metadata such as `archive.rds` or `current.rds` which support I added here, so I renamed asset kind to fit them as well. In existing database there is only one metadata type so assets are marked correctly if it's just a mark and it doesn't affect any other logic that I'm not aware of.

doddi commented:
> > Question: Why have we changed from `PACKAGES` to METADATA`? Do we need to consider existing databases that will already be marked incorrectly as `PACKAGES`?
> 
> Before this pull request R format supported only `PACKAGES.gz` metadata. But there are plenty of other metadata such as `archive.rds` or `current.rds` which support I added here, so I renamed asset kind to fit them as well. In existing database there is only one metadata type so assets are marked correctly if it's just a mark and it doesn't affect any other logic that I'm not aware of.

If an existing nxrm already has R community of a provious version with populated assets then any comparisons in code will be looking for the new `METADATA` but existing assets would be stored as `PACKAGES`. I am thinking this is not an issue because we are not supporting any previous versions of R. However, it seems like a breaking changing when it doesnt have to be.

chernikovmaksym commented:
> > > Question: Why have we changed from `PACKAGES` to METADATA`? Do we need to consider existing databases that will already be marked incorrectly as `PACKAGES`?
> > 
> > 
> > Before this pull request R format supported only `PACKAGES.gz` metadata. But there are plenty of other metadata such as `archive.rds` or `current.rds` which support I added here, so I renamed asset kind to fit them as well. In existing database there is only one metadata type so assets are marked correctly if it's just a mark and it doesn't affect any other logic that I'm not aware of.
> 
> If an existing nxrm already has R community of a provious version with populated assets then any comparisons in code will be looking for the new `METADATA` but existing assets would be stored as `PACKAGES`. I am thinking this is not an issue because we are not supporting any previous versions of R. However, it seems like a breaking changing when it doesnt have to be.

In this request I'm adding support of different metadata. It seems to be a kind of breaking changing to me. Leaving asset kind name `PACKAGES` and add another one for all other metadata is one of the ways. But `PACKAGES.rds`, which is also a package file but with another extension and compression, will be marked as `METADATA` (cause I have generic routing for .rds files) and I think it will be confusing. Creating routings for all known packages seems to be overkill to me. Let me know if you have better option.

chernikovmaksym commented:
@doddi I split `.rds` metadata and `PACKAGES` into different asset kinds and attached them to different routings. Please, review changes here: 338fa1831974c5a79d2d7e4a69d5a5f0c45e894b

doddi commented on nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RHostedRecipe.groovy:
```diff
@@ -73,7 +73,7 @@ class RHostedRecipe
 
     builder.route(packagesGzMatcher()
```
Question: Why is PackagesGz handled differently for hosted implementation?

chernikovmaksym replied:
Cause it's the only metadata hosted supports for now. PACKAGES.gz is automatically generated in facet for hosted packages. I already created ticket to implement other ones.

chernikovmaksym commented:
> Maybe not for this PR but I noticed that hosted basically allows anything to be uploaded. This seems dangerous

Yep, I'll discuss it with the team and create ticket if it doesn't exist.

chernikovmaksym commented:
> lgtm - I would like the upload * addressing before release

Created ticket for it
